draft-ietf-ppm-dap.md

```diff
@@ -1096,84 +1096,169 @@ refinement process. We instead think of these as properties of the output
-      {} - Indicates a field that is only sent if the VDAF
-           requires at least two more rounds.
-       I - Produce first-round prep-share.
-       P - Combine prep shares into prep message.
-       N - Produce next prep share.
-       F - Produce output share.
+In general, reports cannot be aggregated until the Collector specifies an
+aggregation parameter. However, in some situations it is possible to begin
+aggregation as soon as reports arrive. For example, Prio3 has just one valid
+aggregation parameter (the empty string). And there are use cases for Poplar1
+in which aggregation can begin immediately (i.e., those for which the candidate
+prefixes/strings are fixed in advance).
+
+The aggregation flow can be thought of as having three phases:
+
+- Initialization: Begin the aggregation flow by disseminating report shares and
+  initializing the VDAF preparation state for each report.
+- Continuation: Continue the aggregation flow by exchanging prep shares and
+  messages until preparation completes or an error occurs.
+- Completion: Finish the aggregate flow, yielding an output share corresponding
+  to each report report share in the batch.
+
+These phases are described in {{agg-init}} and {{agg-continue-flow}}.
+
+### Preparation Messages
+
+[TODO: Replace definitions of ReportShareError, PrepareStepState, and
+PrepareStep with backward references to this section.]
+
+At any point in the preparation process, a report can be rejected by one of the
+Aggregators and removed from the aggregation job. When a report is rejected, it
+is replaced by a `ReportShareError`:
+
+~~~ message
+enum {
+  batch_collected(0),
+  report_replayed(1),
+  report_dropped(2),
+  hpke_unknown_config_id(3),
+  hpke_decrypt_error(4),
+  vdaf_prep_error(5),
+  batch_saturated(6),
+  task_expired(7),
+  unrecognized_message(8),
+  report_too_early(9),
+  (255)
+} ReportShareError;
+~~~
+
+As shown in Figure 6 of {{!VDAF}}, the preparation process is organized into a
+sequence of rounds. At the start of each round, each Aggregator produces a
+value called a "preparation message-share", or "prep share" for short. The prep
+shares are combined into the preparation message, or "prep message", which is
+used as input for the next round. (The value output in the final round is the
+Aggregator's output share.)
+
+Each step of the aggregation sub-protocol is encoded as a `PrepareStep` message
+as follows:
+
+~~~ message
+enum {
+  initialize(0),
+  continue(1),
+  finish(2),
+  finished(3),
+  reject(4),
+  (255)
+} PrepareStepState;
+
+struct {
+  PrepareStepState prepare_step_state;
+  select (PrepareStep.prepare_step_state) {
+    case initialize:
+      opaque prep_share<0..2^32-1>;
+    case continue:
+      opaque prep_msg<0..2^32-1>;
+      opaque prep_share<0..2^32-1>;
+    case finish:
+      opaque prep_msg<0..2^32-1>;
+    case finished:
+      Empty;
+    case reject:
+      ReportShareError report_share_error;
+  };
+} PrepareStep;
```

Comment:
@branlwyd: Please weigh in on weather there is a clean way to integrate this structure into #393. I think we definitely not avoiding repeating report IDs, either to avoid sending extra bits or having an extra validation step (pick your poison!).

Reply:
The value would be the ability to describe the state machine without having to loop in the decryption and early validation steps.